src/main/cpp/src/timezones.cu

```diff
@@ -118,6 +120,183 @@ auto convert_timestamp_tz(column_view const& input,
+      cuda::std::chrono::duration_cast<cudf::duration_s>(timestamp.time_since_epoch()).count());
+    // input epoch seconds
+
+    auto const tz_transitions = cudf::list_device_view{transitions, tz_index};
+    auto const list_size      = tz_transitions.size();
+
+    auto const transition_times_utc = cudf::device_span<int64_t const>(
+      utc_instants.data<int64_t>() + tz_transitions.element_offset(0),
+      static_cast<size_t>(list_size));
+
+    auto const transition_times_tz = cudf::device_span<int64_t const>(
+      tz_instants.data<int64_t>() + tz_transitions.element_offset(0),
+      static_cast<size_t>(list_size));
+
+    // step 1: Binary search on utc to find the correct offset to convert utc to local
+    auto const utc_it = thrust::upper_bound(
+      thrust::seq, transition_times_utc.begin(), transition_times_utc.end(), epoch_seconds_utc);
+    auto const utc_idx =
+      static_cast<size_type>(thrust::distance(transition_times_utc.begin(), utc_it));
+    auto const utc_list_offset = tz_transitions.element_offset(utc_idx - 1);
+    auto const to_local_offset =
+      static_cast<int64_t>(utc_offsets.element<int32_t>(utc_list_offset));
+
+    auto const to_local_offset_duration = cuda::std::chrono::duration_cast<duration_type>(
+      cudf::duration_s{static_cast<int64_t>(to_local_offset)});
+
+    auto const duration_typed = cuda::std::chrono::duration_cast<duration_type>(
+      cudf::duration_us{static_cast<int64_t>(duration)});
+
+    // step 2: add the duration to the local timestamp
+    auto const local_timestamp_res = timestamp + to_local_offset_duration + duration_typed;
+
+    auto const result_epoch_seconds = static_cast<int64_t>(
+      cuda::std::chrono::duration_cast<cudf::duration_s>(local_timestamp_res.time_since_epoch())
+        .count());
+
+    // step 3: Binary search on local to find the correct offset to convert local to utc
+    auto const local_it = thrust::upper_bound(
+      thrust::seq, transition_times_tz.begin(), transition_times_tz.end(), result_epoch_seconds);
+    auto local_idx =
+      static_cast<size_type>(thrust::distance(transition_times_tz.begin(), local_it));
+
+    // In GpuTimeZoneDB, we build the transition list with
+    // (utcInstant, utcInstant + OffsetBefore, OffsetAfter) if it is a overlap.
+    // But in the step 3, we actually need to binary search on the utcInstant + OffsetBefore
+    // to find the correct offset to convert local to utc. To reuse the code, we need to
+    // add a special post processing here, to make sure we get the correct id that
+    // utcInstant + OffsetBefore is larger than the result_epoch_seconds.
+    auto const temp_list_offset = tz_transitions.element_offset(local_idx);
+    auto const temp_offset = static_cast<int64_t>(utc_offsets.element<int32_t>(temp_list_offset));
+
+    // We don't want to check this if the idx is the first or last
+    if (local_idx != 0 && transition_times_utc[local_idx] != INT64_MAX &&
```

Comment:
On a separate note. Why are the offsets stored as a long? This would be a follow on issue just to investigate if we could get away with an int for the offset. We shouldn't have any offsets of more than a day really.The offsets are stored in seconds. Which would need just larger than a short to store 24 hours.  Even java returns the offset in seconds using an int.

https://docs.oracle.com/javase/8/docs/api/java/time/ZoneOffset.html#getTotalSeconds--

I think we probably could get away with changing it and using INT32_MAX here instead.

Reply:
Nit: use std::numeric_limits<Type>::max() instead of INT64_MAX.